samza-core/src/main/java/org/apache/samza/operators/StreamGraphSpec.java

```diff
@@ -69,87 +77,104 @@
    */
   private int nextOpNum = 0;
   private final Set<String> operatorIds = new HashSet<>();
-  private Serde<?> defaultSerde = new KVSerde(new NoOpSerde(), new NoOpSerde());
   private ContextManager contextManager = null;
+  private Optional<SystemDescriptor> defaultSystemDescriptorOptional = Optional.empty();
 
   public StreamGraphSpec(Config config) {
     this.config = config;
   }
 
   @Override
-  public void setDefaultSerde(Serde<?> serde) {
-    Preconditions.checkNotNull(serde, "Default serde must not be null");
+  public void setDefaultSystem(SystemDescriptor<?> defaultSystemDescriptor) {
+    Preconditions.checkNotNull(defaultSystemDescriptor, "Provided defaultSystemDescriptor must not be null.");
+    String defaultSystemName = defaultSystemDescriptor.getSystemName();
     Preconditions.checkState(inputOperators.isEmpty() && outputStreams.isEmpty(),
-        "Default serde must be set before creating any input or output streams.");
-    this.defaultSerde = serde;
+        "Default system must be set before creating any input or output streams.");
+    Preconditions.checkState(!systemDescriptors.containsKey(defaultSystemName) ||
```

Comment:
Minor: Looks like you do this check for default/input/output. Could you extract to a shared method?